content/advent-2017/mphf.md

```diff
@@ -0,0 +1,269 @@
+Unlike the previous algorithm, this one has no issues with large key sets.
+Constructing the hash function for this wordlist takes only 100ms-125ms.
+
+There are three ways to judge a hash function: construction time, evaluation
+time, and space usage. Our first successful hash function had virtually no
+space usage, a fast evaluation time, but a huge construction time.
+
+This algorithm has a pretty small construction cost. It turns out to be
+linear in the number of keys. The evaluation time is also constant time: one
+"standard" hash function evaluation, some integer mixing, and two table
+lookups.
+
+As for space usage, in my implementation this algorithm uses 8 bytes per
+entry: each key gets its own 4-byte index (0..N-1) and another 4-byte seed
+for the second hash function. The paper gives a method that allows the
+intermediate arrays to be compressed to reduce the space needed, but still
+used for querying without decompression. As is, they can easily be written
+out to disk and loaded back later, or even by a different process. The slices
+could even be accessed via `mmap`.
+
+For my version, I could actually reduce the space usage a little bit at the
+cost of a performance hit. In order to make the lookups faster, the arrays
+are sized to be the next larger power of two. When we hash, we use a bitmask
+to get the appropriate slot in the table. If we replace the bit-mask with a
+(much slower) modulo operator, then we could properly size the arrays with
+exactly N entries.
+
+*Massive Key Sets*
+
+Using 8 bytes per entry might not seem like much, but what if you have a
+billion keys? Or 10 billion?
+
+Last February I saw a paper [Fast and scalable minimal perfect hashing for
+massive key sets](https://arxiv.org/abs/1603.04330).
+
+This paper aims not only to be a fast construction of a minimal perfect hash
+function, but also to drastically reduce the space needed to store the
+mappings.
+
+Similar to the two-level hashing used for hash/displace, this algorithm uses
+multiple hash functions to deal with collisions. However, instead of the
+targets being hash table entries, the targets are *bits* in a bit vector. If
+only a single key hashes to a particular bit, then the bit is set to 1. If more
+that one key hashes to that bit, then the bit is left as 0 and the keys that
+collide are moved to the next layer down. Eventually, all the keys have been
+placed at some level. To look up a value, we must find out which bit it maps
+to. We hash the key with the first hash function and look up that bit in the
+first-level bitvector. If it's a 1, we stop. If it's a 0, we move to searching
+in the second-level bitvector with the second hash function, and so on. In
+order to figure out the value 0..N-1 to return for the hash function, the
+algorithm uses a trick common in [succinct data structures]
+(https://en.wikipedia.org/wiki/Succinct_data_structure). We know there must be
```

Comment:
Not sure what template engine this will get run through, but GH markdown splits the link across the newline.